engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoPolyglotJava.java

```diff
@@ -152,58 +155,86 @@ static void close(EnsoContext ctx) {
-    } catch (InteropException ex) {
-      logger.log(Level.WARNING, "Cannot register findLibraries", ex);
+      if (polyglotJava != this) {
+        return polyglotJava;
+      }
+      polyglotJava = createPolyglotJava(ctx);
+      while (!pendingPath.isEmpty()) {
+        InteropLibrary.getUncached()
+            .invokeMember(polyglotJava, "addPath", pendingPath.remove(0).toString());
+      }
+      try {
+        InteropLibrary.getUncached()
+            .invokeMember(polyglotJava, "findLibraries", new LibraryResolver());
+      } catch (InteropException ex) {
+        logger.log(Level.WARNING, "Cannot register findLibraries", ex);
+      }
+      return polyglotJava;
+    } finally {
+      lock.release();
     }
-    return polyglotJava;
   }
 
   /**
    * This method ensure that hosted as well as guest classpath is the same. This is necessary until
    * real isolation between libraries is implemented.
    */
-  static void addToClassPath(EnsoContext ctx, Object whoIsIgnored, File path)
+  static void addToClassPath(
+      EnsoContext ctx, Object whoIsIgnored, File path, boolean polyglotContextEntered)
       throws InteropException {
     var data = KEY.get(ctx);
-    data.hosted.addToClassPath(path);
-    data.guest.addToClassPath(path);
+    data.hosted.addToClassPath(path, polyglotContextEntered);
+    data.guest.addToClassPath(path, polyglotContextEntered);
   }
 
   /**
    * Modifies the classpath to use to lookup {@code polyglot java} imports.
    *
    * @param file the file to register
+   * @param polyglotContextEntered if true, any lock acquisition will be interruptable for Truffle's
+   *     Safepoints purposes
    */
   @CompilerDirectives.TruffleBoundary
-  private final synchronized void addToClassPath(File file) throws InteropException {
-    if (polyglotJava == this) {
-      pendingPath.add(file);
+  private final void addToClassPath(File file, boolean polyglotContextEntered)
+      throws InteropException {
+    if (polyglotContextEntered) {
+      TruffleSafepoint.setBlockedThreadInterruptible(null, Semaphore::acquire, lock);
     } else {
-      InteropLibrary.getUncached().invokeMember(polyglotJava, "addPath", file.toString());
+      try {
+        lock.acquire();
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+    }
+    try {
+      if (polyglotJava == this) {
+        pendingPath.add(file);
+      } else {
+        InteropLibrary.getUncached().invokeMember(polyglotJava, "addPath", file.toString());
+      }
+    } finally {
+      lock.release();
     }
   }
 
-  private final synchronized void close() {
-    if (polyglotJava instanceof TruffleObject closeJava) {
-      polyglotJava = null;
-      try {
-        InteropLibrary.getUncached().invokeMember(closeJava, "close");
-      } catch (InteropException ex) {
-        logger.log(Level.WARNING, "Cannot close " + closeJava, ex);
+  private final void close() {
```

Comment:
```suggestion
  @TruffleBoundary
  private final void close() {
```
- trying it in #14238

Reply:
- #14238 seems green!